recipes/mpfr/all/conanfile.py

```diff
@@ -20,155 +26,174 @@ class MpfrConan(ConanFile):
-        return self._cmake
+            tc.extra_cflags.append(f"-I{self.build_folder}")
+        if is_msvc(self):
+            tc.extra_cflags.append("-FS")
+
+        env = tc.environment()
+        if is_msvc(self):
+            env.define("CC", "cl -nologo")
+            env.define("CXX", "cl -nologo")
+        tc.generate(env) # Create conanbuild.conf
 
     def _extract_makefile_variable(self, makefile, variable):
-        makefile_contents = tools.load(makefile)
-        match = re.search("{}[ \t]*=[ \t]*((?:(?:[a-zA-Z0-9 \t.=/_-])|(?:\\\\\"))*(?:\\\\\n(?:(?:[a-zA-Z0-9 \t.=/_-])|(?:\\\"))*)*)\n".format(variable), makefile_contents)
+        makefile_contents = load(self, makefile)
+        match = re.search(f'{variable}[ \t]*=[ \t]*((?:(?:[a-zA-Z0-9 \t.=/_-])|(?:\\\\\"))*(?:\\\\\n(?:(?:[a-zA-Z0-9 \t.=/_-])|(?:\\\"))*)*)\n', makefile_contents)
         if not match:
-            raise ConanException("Cannot extract variable {} from {}".format(variable, makefile_contents))
+            raise ConanException(f"Cannot extract variable {variable} from {makefile_contents}")
         lines = [line.strip(" \t\\") for line in match.group(1).split()]
         return [item for line in lines for item in shlex.split(line) if item]
 
     def _extract_mpfr_autotools_variables(self):
-        makefile_am = os.path.join(self._source_subfolder, "src", "Makefile.am")
+        makefile_am = os.path.join(self.source_folder, "src", "Makefile.am") # src/src/Makefile.am
         makefile = os.path.join("src", "Makefile")
         sources = self._extract_makefile_variable(makefile_am, "libmpfr_la_SOURCES")
         headers = self._extract_makefile_variable(makefile_am, "include_HEADERS")
         defs = self._extract_makefile_variable(makefile, "DEFS")
         return sources, headers, defs
 
-    @contextlib.contextmanager
-    def _build_context(self):
-        if self.settings.compiler == "Visual Studio":
-            with tools.vcvars(self):
-                env = {
-                    "AR": "lib",
-                    "CC": "cl -nologo",
-                    "CXX": "cl -nologo",
-                    "LD": "link",
-                    "NM": "dumpbin -symbols",
-                    "OBJDUMP": ":",
-                    "RANLIB": ":",
-                    "STRIP": ":",
-                }
-                with tools.environment_append(env):
-                    yield
-        else:
-            yield
-
     def build(self):
-        for patch in self.conan_data.get("patches", {}).get(self.version, []):
-            tools.patch(**patch)
+        apply_conandata_patches(self)
+
+        if self.options.autogen:
+            # Only apply this patch when building from Git source to prevent Makefile from
+            # invoking autoconf due to to changed .ac or .m4 file
+            patch(self, patch_file=os.path.join(self.export_sources_folder, "patches", f'{self.version}-0001-configure.ac-fixes.patch'))
```

Comment:
Please remove this logic and the subsequent autogen

Reply:
Removed, but please see message above about how an enhancement to apply_conandata_patches would have made this cleaner.